File: sudoku/solver.py
import copy
import csv
import math
import os
import time

from os import path

N = 9
C = math.floor(math.sqrt(N))


current_directory = os.getcwd()
test_validation_directory = path.join(current_directory, "data", "validation")
test_data_directory = path.join(current_directory, "data", "cases")

def read_csv(file_path):
    file = open(file_path, encoding="utf-8-sig", mode="r")

    test_data = file.read().splitlines()
    reader = csv.reader(test_data, delimiter=",")
    file.close()

    rows = []
    for row in reader: 
        rows.append(row)
    return rows

def parse_string_data(line):
    grid = [['' for i in range(N)] for j in range(N)]
    for idx in range(N**2):
        try:
            y = math.floor(idx / N)
            x = idx % N
            grid[y][x] = line[idx]
        except:
            continue
    return grid


# 2-dimensions
# 0 = row
# 1 = column
def valid_digit(cell):
    try:
        cell_value = int(cell)
        return (0 < cell_value and cell_value <= N)
    except:
        return False

def get_row_digits(grid=[[]], row=0):
    return grid[row]
def get_column_digits(grid=[[]], column=0):
    return [grid[i][column] for i in range(N)]
def get_grid_digits(grid=[[]], y=0, x=0):
    return [grid[i + (x * C)][j + (y * C)] for i in range(C) for j in range(C)]

def validate_row(grid=[[]], row=0):
    digits = get_row_digits(grid, row)
    missing_digits = find_missing_digits(digits)
    return len(missing_digits) == 0

def validate_column(grid=[[]], column=0):
    digits = get_column_digits(grid, column)
    missing_digits = find_missing_digits(digits)
    return len(missing_digits) == 0

def validate_grid(grid=[[]], x=0, y=0):
    digits = get_grid_digits(grid, x, y)
    missing_digits = find_missing_digits(digits)
    return len(missing_digits) == 0

def find_missing_digits(digits=[]):
    index_list = [-1 for x in range(N + 1)]
    for index, number in enumerate(digits):
        try:
            digit = int(number)
            index_list[digit] = index
        except:
            #print("Invalid input:", number)
            continue

    missing_numbers = []
    for i in range(1, N + 1):
        if (index_list[i] == -1):
            missing_numbers.append(i)
    return missing_numbers


def sudoku_validate(grid=[[]]):
    invalid_rows = []
    invalid_columns = []
    invalid_grids = []
    for y in range(0, N):
        if (validate_row(grid, y) != True):
            invalid_rows.append(y + 1)
        if (validate_column(grid, y) != True):
            invalid_columns.append(y + 1)

        if (y < C):
            for x in range(0, C):
                if (validate_grid(grid, y, x) != True):
                    invalid_grids.append([y, x])

    is_valid = True
    if (len(invalid_rows) > 0):
        #print("Invalid rows", invalid_rows)
        is_valid = False
    if (len(invalid_columns) > 0):
        #print("Invalid columns", invalid_columns)
        is_valid = False
    if (len(invalid_grids) > 0):
        #print("Invalid grids", invalid_grids)
        is_valid = False
    return is_valid

def find_cell_options(grid=[[]], y=0, x=0):
    gridy = math.floor(y / C)
    gridx = math.floor(x / C)

    dr = get_row_digits(grid, y)
    dc = get_column_digits(grid, x)
    dg = get_grid_digits(grid, gridx, gridy)

    mr = find_missing_digits(dr)
    mc = find_missing_digits(dc)
    mg = find_missing_digits(dg)
    #print(y, dr, mr)
    #print(x, dc, mc)
    #print(gridy, gridx, dg, mg)
    #print(set(mr + mc + mg), list(set(mr + mc + mg)))
    #return list(set(mr + mc + mg))
    return find_missing_digits(dr + dc + dg)

def sudoku_solve(grid=[[]]):
    empty_cells = []
    for y in range(N):
        for x in range(N):
            if not valid_digit(grid[y][x]):
                empty_cells.append([y, x])
    return sudoku_solve_cell(grid, empty_cells)

def sudoku_solve_cell(grid=[[]], cells=[], idx=0):
    if (idx < len(cells)):
        [y, x] = cells[idx]
        missing_digits = find_cell_options(grid, y, x)
        #print(y, x, missing_digits)
        for z in missing_digits:
            #print(y, x, z)
            grid_step = copy.deepcopy(grid)
            grid_step[y][x] = str(z)
            solution = sudoku_solve_cell(grid_step, cells, idx + 1)
            if (solution != None):
                #print("Solution found")
                return solution
    else:
        is_solution_valid = sudoku_validate(grid)
        #for row in grid:
        #    print(row)
        #print(is_solution_valid)
        if (is_solution_valid):
            return grid
        


print("  Testing positive cases")
positive_case_directory = path.join(test_validation_directory, "positive")
positive_case_files = os.listdir(positive_case_directory)
for file_name in positive_case_files:
    file_path = path.join(positive_case_directory, file_name)
    rows = read_csv(file_path)

    case_name = file_name.split(".")[0]
    validation = sudoku_validate(rows)
    print("Expect test-case, " + case_name + ", to be True. Validation found it to be " + str(validation) + "...")
print()

print("  Testing negative cases")
negative_case_directory = path.join(test_validation_directory, "negative")
negative_case_files = os.listdir(negative_case_directory)
for file_name in negative_case_files:
    file_path = path.join(negative_case_directory, file_name)
    rows = read_csv(file_path)

    case_name = file_name.split(".")[0]
    validation = sudoku_validate(rows)
    print("Expect test-case, " + case_name + ", to be False. Validation found it to be " + str(validation) + "...")
print()

print("  Running cases")
case_files = os.listdir(test_data_directory)
for file_name in case_files:
    print(file_name)
    start_time_ms = int(time.time() * 1000)
    file_path = path.join(test_data_directory, file_name)
    rows = read_csv(file_path)
    solution = sudoku_solve(rows)
    if (solution != None):
        #for row in solution:
        #    print(row)
        end_time_ms = int(time.time() * 1000)
        print("Runtime: " + str(round(end_time_ms - start_time_ms)) + "ms")

'''
file_path = path.join(current_directory, "data", "top1465.txt")
file = open(file_path, mode="r")
test_data = file.read().splitlines()
file.close()
for raw_data in test_data[0:2]:
    print(raw_data)
    rows = parse_string_data(raw_data)
    start_time_ms = int(time.time() * 1000)
    
    for row in rows:
        print(row)
    solution = sudoku_solve(rows)
    print(solution)
    if (solution != None):
        for row in solution:
            print(row)
        end_time_ms = int(time.time() * 1000)
        print("Runtime: " + str(round(end_time_ms - start_time_ms)) + "ms")
'''

File: sudoku/solver.rb
include Math

N = 9
C = sqrt(N).floor


current_directory = Dir::getwd()
test_validation_directory = File.join(current_directory, "data", "validation")
test_data_directory = File.join(current_directory, "data", "cases")


=begin
definition: Reads .csv. file that is delimited by comma
param {String} file_path
returns {Array<Number>}
=end
def read_csv(file_path = "")
	rows = []
	file = File.new(file_path, "r")
	# See String#encode documentation
	encoding_options = {
	  :invalid           => :replace,  # Replace invalid byte sequences
	  :undef             => :replace,  # Replace anything not defined in ASCII
	  :replace           => '',        # Use a blank for those replacements
	  :universal_newline => true       # Always break lines with \n
	}
	while (line = file.gets)
		clean_line = line.split("").select{|i| i.ord != 65279}.join.to_s
		row = clean_line.split(",").map {|s| s.strip}
		rows.push(row)
	end
	file.close

	return rows
end	


=begin
definition: Retrieve digits in a row
param {Array<Array<Number>>} grid
param {Number} row
returns {Array<Number>}
=end
def get_row_digits(grid=[[]], row=0)
	return grid[row]
end
=begin
definition: Retrieve digits in a column
param {Array<Array<Number>>} grid
param {Number} column
returns {Array<Number>}
=end
def get_column_digits(grid=[[]], column=0)
	return grid.map {|row| row[column]}
end
=begin
definition: Retrieve digits in a box
param {Array<Array<Number>>} grid
param {Number} x
param {Number} y
returns {Array<Number>}
=end
def get_box_digits(grid=[[]], x=0, y=0)
	return (0..N-1).map {|i| grid[(i / C).floor + (y * C)][(i % C) + (x * C)] }
=begin
		box_x = (i % C)
		box_y = (i / C).floor
		return grid[box_y + (y * C)][box_x + (x * C)]
	}
=end
end

=begin
definition: Ensure unique digits across row
param {Array<Array<Number>>} grid
param {Number} row
returns {Boolean}
=end
def validate_row(grid=[[]], row=0)
	digits = get_row_digits(grid, row)
	missing_digits = find_missing_digits(digits)
	return missing_digits.length == 0
end
=begin
definition: Ensure unique digits across column
param {Array<Array<Number>>} grid
param {Number} column
returns {Boolean}
=end
def validate_column(grid=[[]], column=0)
	digits = get_column_digits(grid, column)
	missing_digits = find_missing_digits(digits)
	return missing_digits.length == 0
end
=begin
definition: Ensure unique digits across box
param {Array<Array<Number>>} grid
param {Number} x
param {Number} y
returns {Boolean}
=end
def validate_box(grid=[[]], x=0, y=0)
	digits = get_box_digits(grid, x, y)
	missing_digits = find_missing_digits(digits)
	return missing_digits.length == 0
end

=begin
definition: Identify digits that are missing from a set
param {Array<Number>} digits
returns {Array<Number>}
=end
def find_missing_digits(digits=[])
	index_list = {}
	digits.each_with_index do |number, index|
		begin
			digit = number.strip.to_i
			index_list[digit] = index
		rescue
			#puts "Invalid input: " + number
		end
	end

	missing_numbers = []
	return (1..N).select {|number| (not index_list.has_key?(number)) }
end

=begin
definition: Validate sudoku grid solution
			1. Ensure unique digits across row
			2. Ensure unique digits across column
			3. Ensure unique digits across 3x3 box
param {Array<Array<Number>>} grid
returns {Boolean}
=end
def sudoku_validate(grid = [[]])
	invalid_rows = []
	invalid_columns = []
	invalid_boxes = []
	(0..N-1).each do |y|
		#puts "row/column validation", y, validate_row(grid, y), validate_column(grid, y)
		if (validate_row(grid, y) == false)
			invalid_rows.push(y + 1)
		end
		if (validate_column(grid, y) == false)
			invalid_columns.push(y + 1)
		end

		box_x = (y % C)
		box_y = (y / C).floor
		if (validate_box(grid, box_x, box_y) == false)
			invalid_boxes.push([box_x, box_y])
		end
		#puts "box validation", box_x, box_y, validate_box(grid, box_x, box_y)
	end

	is_valid = true
	if (invalid_rows.length > 0)
		#puts "Invalid rows: " + invalid_rows.to_s
		is_valid = false
	end
	if (invalid_columns.length > 0)
		#puts "Invalid columns: " + invalid_columns.to_s
		is_valid = false
	end
	if (invalid_boxes.length > 0)
		#puts "Invalid boxes: " + invalid_boxes.to_s
		is_valid = false
	end
	return is_valid
end

puts "  Testing positive cases"
positive_case_directory = File.join(test_validation_directory, "positive")
Dir.entries(positive_case_directory)
	.select {|file_name| file_name.end_with?(".csv") }
	.each {|file_name|
		file_path = File.join(positive_case_directory, file_name)
		rows = read_csv(file_path)

		case_name = file_name.split(".")[0]
		validation = sudoku_validate(rows)
		puts "Expect test-case, " + case_name + ", to be true. Validation found it to be " + validation.to_s
	}

puts "  Testing negative cases"
negative_case_directory = File.join(test_validation_directory, "negative")
Dir.entries(negative_case_directory)
	.select {|file_name| file_name.end_with?(".csv") }
	.each {|file_name|
		file_path = File.join(negative_case_directory, file_name)
		rows = read_csv(file_path)

		case_name = file_name.split(".")[0]
		validation = sudoku_validate(rows)
		puts "Expect test-case, " + case_name + ", to be false. Validation found it to be " + validation.to_s
	}
